src/ripple/basics/SlabAllocator.h

```diff
@@ -30,67 +30,176 @@
+    struct alignas(Type) FreeItem
+    {
+        FreeItem* next;
+
+        FreeItem(FreeItem* n = nullptr) noexcept : next(n)
+        {
+        }
+
+        ~FreeItem() noexcept = default;
+    };
+
+    static_assert(sizeof(Type) >= sizeof(FreeItem));
+
     /** A block of memory that is owned by a slab allocator */
     struct SlabBlock
     {
-        // A mutex to protect the freelist for this block:
+        // A mutex to protect this block:
         std::mutex m_;
 
         // A linked list of appropriately sized free buffers:
-        std::uint8_t* l_ = nullptr;
+        FreeItem* list_ = nullptr;
 
-        // The next memory block
+        // The next block:
         SlabBlock* next_;
 
         // The underlying memory block:
-        std::uint8_t const* const p_ = nullptr;
+        std::uint8_t* data_ = nullptr;
+
+        // The number of items available in this block
+        std::size_t avail_{0};
+
+        explicit SlabBlock(SlabBlock* next) : next_(next)
+        {
+        }
 
-        // The extent of the underlying memory block:
-        std::size_t const size_;
+        // We don't release the allocated memory here: slabs are only
+        // destroyed when the process is shutting down, and we cannot
+        // guarantee that this isn't called before the destruction of
+        // all the objects that are allocated in the buffer we have.
+        ~SlabBlock() = default;
 
-        SlabBlock(
-            SlabBlock* next,
+        std::uint8_t*
+        assign_and_allocate(
             std::uint8_t* data,
             std::size_t size,
-            std::size_t item)
-            : next_(next), p_(data), size_(size)
+            std::size_t item,
+            std::size_t count)
         {
-            // We don't need to grab the mutex here, since we're the only
-            // ones with access at this moment.
+            assert(
+                reinterpret_cast<std::uintptr_t>(data) %
+                    alignof(decltype(data_)) ==
+                0);
+
+            assert(count == (size / item));
+
+            assert(data != nullptr);
+
+            std::lock_guard _(m_);
+
+            assert(
+                (avail_ == 0 && list_ == nullptr) ||
+                (avail_ != 0 && list_ != nullptr));
 
-            while (data + item <= p_ + size_)
+            // It's possible that this block has an underlying buffer. If
+            // it does, and it has space available, we use it.
+            if (data_ != nullptr)
             {
-                // Use memcpy to avoid unaligned UB
-                // (will optimize to equivalent code)
-                std::memcpy(data, &l_, sizeof(std::uint8_t*));
-                l_ = data;
-                data += item;
+                auto ret = list_;
+
+                if (ret != nullptr)
+                {
+                    list_ = ret->next;
+                    --avail_;
+                }
+
+                return reinterpret_cast<std::uint8_t*>(ret);
```

Comment:
We _may_ need `std::launder` here. I'll admit I don't actually know for sure, I'm trying to read the `std::launder` docs to know for sure.

Reply:
I don't think we do, but would welcome more conversation on the topic.

Given:

```
template< class T >
constexpr T* launder( T* p ) noexcept;
```

The documentation states:

> Formally, given
>
>* the pointer `p` represents the address `A` of a byte in memory
>* an object `x` is located at the address `A`
>* `x` is within its [lifetime](https://en.cppreference.com/w/cpp/language/lifetime)
>* the type of `x` is the same as `T`, ignoring cv-qualifiers at every level
>* every byte that would be reachable through the result is reachable through `p` (bytes are reachable through a pointer that points to an object `y` if those bytes are within the storage of an object `z` that is [pointer-interconvertible](https://en.cppreference.com/w/cpp/language/static_cast#pointer-interconvertible) with `y`, or within the immediately enclosing array of which `z` is an element). 
>
>Then `std::launder(p)` returns a value of type `T*` that points to the object `x`.

I'll admit that I find all this _really_ confusing, and wish someone would do the ELI5 thing.

But, my read is that in this case `p` is a `FreeItem`, and so `std::launder` can only give us _that_ back. Also, a `FreeItem` is likely to be smaller in size than whatever the new memory is used for by the caller, so I'm not sure that the reachability analysis is satisfied either.

Given this description of a use case from the docs: "[o]btaining a pointer to an object created by placement new from a pointer to an object providing storage for that object" I do think that there's an argument to be made that whoever invokes placement new to construct an object in the returned buffer should pass the result of `std::launder` but at the same time I'm confused about how to do that without violating one of the above requirements and, therefore, invoking UB.

Frankly, I think that, as is, `std::launder` is nigh unusable.